### DISCUSSION TITLE
Additional Assertion AssertSetContains and Inverse

### DISCUSSION THREAD

**Discussion by humanatepixel** [livewire/livewire#8408] [3. Feature ideas]
Before I go and write the PR which adds the assertion, I just wanted to check. Do we agree that there should be an `assertSetContains(string $prop, array $needles)` and it's antithesis?

Maybe there's already something that does it, but I struggle to `get` props in a livewire test, because everythings `static`

A quick workaround would probably be something like 
```
Livewire::test(Component::class)
  ->assertCallback(fn ($comp) => {
  // assert the prop contains whatever here
});

If there's something like assertCallback then I'll just use that, but couldn't see it in docs or the actual library code.

Note: I'm new to livewire, but not TDD

Other Note: While writing this feature idea, I had the idea for that `assertCallback` so I'm happy to do either one as a PR if someone thinks it has legs.

**Comment by joshhanley** [comment]
@humanatepixel how would that work in comparison to `assertSet`? Can you provide some examples (code) of how you would use it?

**Reply by humanatepixel** [reply]
Sure


```
<?php
// Todoer.php

namespace App\Livewire\Todo;

use Livewire\Component;

class Todoer extends Component
{
    public array $todos;
    
    public function mount() {
        $this->todos = ['test', 'this', 'thing'];
    }
    public function changeTodos($value)
    {
        $this->todos = array_merge($this->todos, [$value]);
    }     
}
```

```
<?php
// tests/Feature/Livewire/Todo/TodoerTest.php

namespace Tests\Feature\Livewire\Todo;

use Illuminate\Foundation\Testing\TestCase;
use Livewire\Livewire;

class TodoerTest extends TestCase
{
    public function test_todoer_changed() {
        $newValue = 'unit-testing';

        // Assert unchanged state for demo
        Livewire::test(Todoer::class)
            ->assertSetContains('todos', ['test']);

        // Change it for demo
        // Assert unchanged state for demo
        Livewire::test(Todoer::class)
            ->call('changeTodos', $newValue)
            ->assertSetContains('todos', ['thing', $newValue]);

    }

    public function test_todoer_changed_with_callback() {
        $newValue = 'unit-testing';

        // Assert unchanged state for demo
        Livewire::test(Todoer::class)
            ->assertCallback(fn($todoer) => in_array('test', $todoer->todos));

        // Change it for demo
        // Assert unchanged state for demo
        Livewire::test(Todoer::class)
            ->call('changeTodos', $newValue)
            ->assertCallback(fn($todoer) => in_array('test', $todoer->todos) && in_array($newValue, $todoer->todos));

    }
}
```
Something like that. I think the main issue I run into when testing livewire components is we can set props, but can't get them. I'd like to be able to assert the state of props with a bit more flexibility than we currently have.

@joshhanley thanks for keeping the conversation going

**Reply by humanatepixel** [reply]
I skipped the rendering and things necessary to actually run the code because we're in theorising mode ;)

**Reply by humanatepixel** [reply]
Additional thinking, you can probably use `assertSet` but that requires you to have the full expected result for the result, as it uses PHPUnit::assertSame or PHPUnit::assertEquals under the hood

If we don't do the `assertCallback` we'd probably just use `PHPUnit::assertContains`

**Comment by humanatepixel** [comment]
There is a workaround, I didn't realise you could get the component instance from `Livewire::test()`
```
        $todoer = Livewire::test(Todoer::class)
            ->call('changeTodos', $newValue);
        $this->assertContains($newValue, $todoer->get('todos'));
```
So this is really a feature request around being able to chain with the niceness of `assertContains`, and maybe that helps with Pest (I haven't looked at pest yet)

I'll leave it open if someone wants to consider the idea still

**Comment by humanatepixel** [comment]
@joshhanley (sorry for the ping but) is Discussions the right place to decide this type of quality of life improvement? I can just built the PR, but don't want to unless it's > 80% likely to be merged (been bit with open source communities)

[I will say, generally with the laravel dev community it has been a welcome breath of fresh air!]

**Reply by joshhanley** [reply]
@humanatepixel no worries! So ultimately Caleb is the one who decides what gets merged/ not merged, so I can't guarantee anything. Discussions are a good place to gauge community interest first though, as that is something Caleb likes to see. But if you want his opinion on it, the best thing is a PR.

But one thing I would do, is first create the assertions as macros (on the `Livewire\Features\SupportTesting\Testable` class) in your own app, make sure they all work happily (plus you get the benefit of using them). Then if you wish to PR it, it's a matter of copying the implementation across to a PR, so won't be wasted effort even if it doesn't get merged as you will already have it 🙂

I have a bunch of macro'd test assertions in my apps for things like `assertCollectionContains` etc.

In terms of the idea, makes sense to me (I misunderstood it when I looked at it yesterday). I would probably call it `assertArrayContains` or something like that.

**Comment by humanatepixel** [comment]
@joshhanley (sorry didn't get notified for some reason)
> But one thing I would do, is first create the assertions as macros (on the `Livewire\Features\SupportTesting\Testable` class) in your own app, make sure they all work happily (plus you get the benefit of using them). Then if you wish to PR it, it's a matter of copying the implementation across to a PR, so won't be wasted effort even if it doesn't get merged as you will already have it 🙂
> 
> I have a bunch of macro'd test assertions in my apps for things like `assertCollectionContains` etc.

Ok yeah, that makes sense. I'll look into the assertion as macros piece and if I get it working in a way that's reusable I can make a PR if I'm feeling generous ;)

Cheers man
